src/herder/TxQueueLimiter.cpp

```diff
@@ -0,0 +1,160 @@
+    auto txGroupIt = mTxGroupsByBaseFee.find(baseFee);
+    if (txGroupIt == mTxGroupsByBaseFee.end())
+    {
+        throw std::logic_error(
+            "invalid state (basefee) removing tx in TxQueueLimiter");
+    }
+    auto& txGroup = txGroupIt->second;
+    auto& txs = txGroup.mTxs;
+    auto it = txs.find(tx);
+    if (it == txs.end())
+    {
+        throw std::logic_error("invalid state removing tx in TxQueueLimiter");
+    }
+    txs.erase(it);
+    if (txs.empty())
+    {
+        mTxGroupsByBaseFee.erase(txGroupIt);
+    }
+    else
+    {
+        // update ops
+        txGroup.mOpsCount -= txOps;
+        // update id tracker
+        auto idTrackerIt = txGroup.mSourceIDToTxCount.find(tx->getSourceID());
+        if (idTrackerIt == txGroup.mSourceIDToTxCount.end())
+        {
+            throw std::logic_error(
+                "invalid state (id tracker) removing tx in TxQueueLimiter");
+        }
+        if (--(idTrackerIt->second) == 0)
+        {
+            txGroup.mSourceIDToTxCount.erase(idTrackerIt);
+        }
+    }
+}
+
+bool
+TxQueueLimiter::canAddTx(TransactionFrameBasePtr& newTx,
+                         TransactionFrameBasePtr& oldTx)
+{
+    auto newOps = mQueueSizeOps + newTx->getNumOperations();
+    if (oldTx)
+    {
+        newOps -= oldTx->getNumOperations();
+    }
+    if (newOps <= maxQueueSizeOps())
+    {
+        return true;
+    }
+    auto txBaseFee = newTx->getBaseFeeBid();
+    // need to see if we could be added by kicking out cheaper transactions
+    // starting with the cheapest group
+    auto neededOps = newOps - maxQueueSizeOps();
+    auto id = newTx->getSourceID();
+    for (auto it = mTxGroupsByBaseFee.begin();
+         it != mTxGroupsByBaseFee.end() && txBaseFee > it->first; ++it)
+    {
+        auto& txGroup = it->second;
+
+        // first check that this group does not contain
+        // transactions for this account
+        if (txGroup.mSourceIDToTxCount.find(id) !=
+            txGroup.mSourceIDToTxCount.end())
+        {
+            return false;
+        }
+        if (txGroup.mOpsCount >= neededOps)
+        {
+            return true;
+        }
+        neededOps -= txGroup.mOpsCount;
+    }
+    return false;
+}
+
+TransactionFrameBasePtr
+TxQueueLimiter::getWorstTransaction()
+{
+    auto it = mTxGroupsByBaseFee.begin();
+    if (it == mTxGroupsByBaseFee.end())
+    {
+        throw std::logic_error(
+            "invalid state getting worst tx in TxQueueLimiter");
+    }
+    // as we hash pointers, picking the first one in hash order is
+    // good enough in this context
+    return *(it->second.mTxs.begin());
+}
+
+bool
+TxQueueLimiter::evictTransactions(
+    size_t ops, std::function<void(TransactionFrameBasePtr&)> evict)
+{
+    while (size() + ops > maxQueueSizeOps())
+    {
+        if (size() == 0)
+        {
+            return false;
+        }
+        auto evictTx = getWorstTransaction();
+        evict(evictTx);
```

Comment:
I cannot do this, see the contract in the header file: this is the responsibility of `evict` (which it does by calling `ban`). The lifetime of transactions is determined by `TransactionQueue`.

If I was calling `removeTransaction` here I would cause this class and the main `TransactionQueue` to be out of sync, and `evict` would end up calling `removeTransaction` again which will trip invariants/violate any lifetime expectations on which transactions should be referenced at any given time by the different classes.

Reply:
Didn't read the comment. This is fine since it's documented.